src/main/java/org/opensearch/security/privileges/ResourceAccessEvaluator.java

```diff
@@ -0,0 +1,212 @@
+     * This allows for a standalone authorization flow for users requesting access to resource.
+     * <p>
+     * 0. Creating a resource requires "create" permissions that are checked outside this evaluator.
+     * 1. Owners and admin-certificate users will be granted access automatically.
+     * 2. Even if a user has access to all indices, they will not be able to access a resource that they are not the owner of and is not shared with them.
+     * 3. A user with no index permissions may not be able to create a resource, however, they can modify and delete a resource shared with them at full-access level.
+     *
+     * @param request                         may contain information about the index and the resource being requested
+     * @param action                          the action being requested to be performed on the resource
+     * @param isResourceSharingFeatureEnabled flag to indicate whether this feature is enabled
+     * @param context                         the evaluation context to be used when performing authorization
+     * @param presponse                       the response which tells whether the action is allowed for user, or should the request be checked with another evaluator
+     * @return PrivilegesEvaluatorResponse may be complete if the request is for a resource and authz check was successful, incomplete otherwise
+     */
+    public PrivilegesEvaluatorResponse evaluate(
+        final ActionRequest request,
+        final String action,
+        boolean isResourceSharingFeatureEnabled,
+        final PrivilegesEvaluationContext context,
+        final PrivilegesEvaluatorResponse presponse
+    ) {
+
+        // If feature is disabled we skip evaluation through this evaluator
+        if (!isResourceSharingFeatureEnabled) {
+            return presponse;
+        }
+
+        // Skip evaluation if request is not a DocRequest type
+        if (!(request instanceof DocRequest req)) {
+            return presponse;
+        }
+
+        log.debug("Evaluating resource access");
+
+        // Resource Creation requests must be checked by regular index access evaluator
+        if (req.id() == null) {
+            log.debug("Request id is null, request is of type {}", req.getClass().getName());
+            return presponse;
+        }
+
+        // if requested index is not a resource sharing index, move on to the next evaluator
+        if (!resourceIndices.contains(req.index())) {
+            log.debug("Request index {} is not a protected resource index", req.index());
+            return presponse;
+        }
+
+        final UserSubjectImpl userSubject = (UserSubjectImpl) this.threadContext.getPersistent(
+            ConfigConstants.OPENDISTRO_SECURITY_AUTHENTICATED_USER
+        );
+        final User user = (userSubject == null) ? null : userSubject.getUser();
+
+        if (user == null) {
```

Comment:
I don't think this condition can ever be true. We already have short-circuiting logic here: https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/filter/SecurityFilter.java#L327-L364

There is some [logic](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/filter/SecurityFilter.java#L342-L345) in that block that I am not super familiar with.